lib/redirect-pool.js

```diff
@@ -0,0 +1,153 @@
+const { Readable } = require('stream')
+const { pipeline, request, stream } = require('./agent')
+const util = require('./core/util')
+const Pool = require('./pool')
+
+const kOrigin = Symbol('Origin')
+const kDispatchMethod = Symbol('RedirectPoolDispatchMethod')
+const kRedirectsLeft = Symbol('RedirectPoolRedirectsLeft')
+
+const redirectCodes = [300, 301, 302, 303, 307, 308]
+const defaultMaxRedirections = 10
+
+class RedirectDiscardedResponse extends Readable {
+  constructor(resume) {
+    super({ autoDestroy: true, read: resume })
+  }
+}
+
+class RedirectPool extends Pool {
+  constructor(origin, options) {
+    super(origin, options)
+
+    this[kOrigin] = origin
+  }
+
+  request(opts, callback) {
+    opts[kDispatchMethod] = 'request'
+
+    return super.request(opts, callback)
+  }
+
+  stream(opts, factory, callback) {
+    opts[kDispatchMethod] = 'stream'
+
+    return super.stream(opts, factory, callback)
+  }
+
+  pipeline(opts, handler) {
+    opts[kDispatchMethod] = 'pipeline'
+
+    return super.pipeline(opts, handler)
+  }
+
+  dispatch(opts, handler) {
+    const pool = this
+
+    // Compute the number of left redirects, if needed
+    if (!(kRedirectsLeft in opts)) {
+      this.parseMaxRedirects(opts)
+    }
+
+    const originalOnHeaders = handler.onHeaders
+
+    handler.onHeaders = function onHeadersWithRedirect(statusCode, headers, resume) {
```

Comment:
If it's a different origin, it should go back to the Agent.

Reply:
I don't see your point. I do that in https://github.com/ShogunPanda/undici/blob/main/lib/redirect-pool.js#L83, which basically calls the top level method. Can you please clarify?

Reply:
A `Pool` has only a single destination of protocol/domain/port. If we are changing those in the redirect, we should go back to the `Agent`. As I read the code, this is not working right now nor there is a test for it.

Reply:
I do instead. When I call request in line https://github.com/ShogunPanda/undici/blob/main/lib/redirect-pool.js#L85, is the top `undici.request` which ultimately calls `dispatchFromAgent` and gets a pool for a new origin from the agent.

About the test, there is explicitly one: https://github.com/ShogunPanda/undici/blob/main/test/redirect-request.js#L339 - In this case only the port changes, but it shouldn't matter, right?

Reply:
Oh, I understand now. I think the Pool should receive an instance of its own agent and not pass through the global agent. The current agent vs the global agent could be different.

Reply:
I think that's already the case. At the moment the pool calls the top-level method by sending the original agent back. 
https://github.com/ShogunPanda/undici/blob/main/lib/agent.js#L106

As you can see the agent (which might or might not be the global one) is send to the pool which in turn send it back to `dispatchFromAgent`. So only a single agent is used. Am I wrong?

Reply:
Also, consider that the agent must be set explicitily configured in order to use the `RedirectPool`, so either `request` and similar are called from with a `agent`, or `setGlobalAgent` has been called.

Reply:
Amazing thanks! This work beautifully!